File: solutions/scala/space-age/1/src/main/scala/SpaceAge.scala
object SpaceAge:
  enum Planet:
    case Mercury,
      Venus,
      Earth,
      Mars,
      Jupiter,
      Saturn,
      Uranus,
      Neptune

  def secondsToYears(age: Double): Double = age / 31_557_600

  def convertAge(age: Double, planet: Planet): Double =
    var conversionRate = planet match {
      case Planet.Mercury => 0.2408467
      case Planet.Venus   => 0.61519726
      case Planet.Earth   => 1.0
      case Planet.Mars    => 1.8808158
      case Planet.Jupiter => 11.862615
      case Planet.Saturn  => 29.447498
      case Planet.Uranus  => 84.016846
      case Planet.Neptune => 164.79132
    }
    secondsToYears(age) / conversionRate

  def onEarth(age: Double) = convertAge(age, Planet.Earth)
  def onMercury(age: Double) = convertAge(age, Planet.Mercury)
  def onVenus(age: Double) = convertAge(age, Planet.Venus)
  def onMars(age: Double) = convertAge(age, Planet.Mars)
  def onJupiter(age: Double) = convertAge(age, Planet.Jupiter)
  def onSaturn(age: Double) = convertAge(age, Planet.Saturn)
  def onUranus(age: Double) = convertAge(age, Planet.Uranus)
  def onNeptune(age: Double) = convertAge(age, Planet.Neptune)

File: solutions/ocaml/space-age/1/space_age.ml
open Base

type planet =
  | Mercury
  | Venus
  | Earth
  | Mars
  | Jupiter
  | Saturn
  | Neptune
  | Uranus

let conversion_rate = function
  | Mercury -> 0.2408467
  | Venus -> 0.61519726
  | Earth -> 1.0
  | Mars -> 1.8808158
  | Jupiter -> 11.862615
  | Saturn -> 29.447498
  | Uranus -> 84.016846
  | Neptune -> 164.79132

let age_on planet age_seconds =
  let earth_years age = Float.of_int age /. 31_557_600.0 in
  earth_years age_seconds /. conversion_rate planet
